nogo exclude_files doesn't work properly for cgo

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What version of rules_go are you using?

<!-- Check io_bazel_rules_go in WORKSPACE if you're not sure -->

v0.27.0

### What version of gazelle are you using?

v0.23.0

### What version of Bazel are you using?

4.0.0

### Does this issue reproduce with the latest releases of all the above?

Yes

### What operating system and processor architecture are you using?

darwin/amd64

### Any other potentially useful information about your toolchain?

<!-- C/C++ compiler, custom CROSSTOOL, remote execution? -->

Nothing special. Using Command Line Tools rather than Xcode.

### What did you do?

<!--
If possible, provide a minimal recipe for reproducing the error.
-->

Excluding files from nogo analysis with the `exclude_files` config key. The filename in question is something like `external/com_blah/common/abc.go` (a file in a Gazelle `go_repository`), and I added `"external/": ""` to the nogo config file.

### What did you expect to see?

Diagnostics from the specified analyzers get excluded when running through this file.

### What did you see instead?

The diagnostics still appear:

```
external/com_blah/common/abc.go:14:23: possible misuse of unsafe.Pointer (unsafeptr)
Target @com_blah//common:common failed to build
```

----

I dug around a bit, and added 

```go
fmt.Fprintf(os.Stderr, "%s(%s)\n", act.a.Name, filename);
```

to right before [line 422 of nogo\_main.go](https://github.com/bazelbuild/rules_go/blob/707c634bf8f7ca5b57c2a923bec16fa62ced1fc7/go/tools/builders/nogo_main.go#L422). Here's what I got in the output logs:

```
compilepkg: unsafeptr(/var/folders/y4/d2xhlc550nv5wxfl3w1mr47xrjmbp2/T/rules_go_work-234795290/abc.cgo1.go)
nogo: errors found by nogo during build-time code analysis:
external/com_blah/common/abc.go:14:23: possible misuse of unsafe.Pointer (unsafeptr)
Target @com_blah//common:common failed to build
```

It appears that this may be a result of using `token.FileSet.File` on [line 397](https://github.com/bazelbuild/rules_go/blob/707c634bf8f7ca5b57c2a923bec16fa62ced1fc7/go/tools/builders/nogo_main.go#L397), which doesn't respect cgo's `//line` adjustment comments, rather than `token.FileSet.Position`.

--- a/go/tools/builders/nogo_main.go
+++ b/go/tools/builders/nogo_main.go
@@ -394,10 +394,10 @@ func checkAnalysisResults(actions []*action, pkg *goPackage) string {
 		for _, d := range act.diagnostics {
 			// NOTE(golang.org/issue/31008): nilness does not set positions,
 			// so don't assume the position is valid.
-			f := pkg.fset.File(d.Pos)
+			p := pkg.fset.Position(d.Pos)
 			filename := "-"
-			if f != nil {
-				filename = f.Name()
+			if p.IsValid() {
+				filename = p.Filename
 			}
 			include := true
 			if len(config.onlyFiles) > 0 {

--- a/tests/core/nogo/custom/custom_test.go
+++ b/tests/core/nogo/custom/custom_test.go
@@ -79,6 +79,13 @@ go_library(
 )
 
 go_library(
+    name = "has_errors_linedirective",
+    srcs = ["has_errors_linedirective.go"],
+    importpath = "haserrors_linedirective",
+    deps = [":dep"],
+)
+
+go_library(
     name = "no_errors",
     srcs = ["no_errors.go"],
     importpath = "noerrors",
@@ -290,8 +297,25 @@ import (
 )
 
 func Foo() bool { // This should fail foofuncname
-	dep.D()     // This should fail visibility
-	return true // This should fail boolreturn
+	dep.D() // This should fail visibility
+	return true
+}
+
+-- has_errors_linedirective.go --
+//line linedirective.go:1
+package haserrors_linedirective
+
+import (
+	/*line linedirective_importfmt.go:4*/ _ "fmt" // This should fail importfmt
+
+	"dep"
+)
+
+//line linedirective_foofuncname.go:9
+func Foo() bool { // This should fail foofuncname
+//line linedirective_visibility.go:10
+	dep.D() // This should fail visibility
+	return true
 }
 
 -- no_errors.go --
@@ -332,7 +356,17 @@ func Test(t *testing.T) {
 				`has_errors.go:.*function D is not visible in this package \(visibility\)`,
 			},
 		}, {
+			desc:        "default_config_linedirective",
+			target:      "//:has_errors_linedirective",
+			wantSuccess: false,
+			includes: []string{
+				`linedirective_importfmt.go:.*package fmt must not be imported \(importfmt\)`,
+				`linedirective_foofuncname.go:.*function must not be named Foo \(foofuncname\)`,
+				`linedirective_visibility.go:.*function D is not visible in this package \(visibility\)`,
+			},
+		}, {
 			desc:        "custom_config",
+			config:      "config.json",
 			target:      "//:has_errors",
 			wantSuccess: false,
 			includes: []string{
@@ -340,7 +374,19 @@ func Test(t *testing.T) {
 				`has_errors.go:.*function must not be named Foo \(foofuncname\)`,
 			},
 			excludes: []string{
-				"custom/has_errors.go:.*function D is not visible in this package",
+				`visib`,
+			},
+		}, {
+			desc:        "custom_config_linedirective",
+			config:      "config.json",
+			target:      "//:has_errors_linedirective",
+			wantSuccess: false,
+			includes: []string{
+				`linedirective_foofuncname.go:.*function must not be named Foo \(foofuncname\)`,
+				`linedirective_visibility.go:.*function D is not visible in this package \(visibility\)`,
+			},
+			excludes: []string{
+				`importfmt`,
 			},
 		}, {
 			desc:        "no_errors",

